extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -6,101 +6,86 @@
+		this._openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		this.books = [];
+		this.initialize(workspaceFolders, null);
+		vscode.commands.executeCommand('setContext', 'untitledBooks', openAsUntitled);
+		this.viewId = view;
+		this.prompter = new CodeAdapter();
 
-	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext) {
-		this.initialize(workspaceFolders, null, extensionContext);
 	}
 
-	private initialize(workspaceFolders: vscode.WorkspaceFolder[], bookPath: string, context: vscode.ExtensionContext): void {
-		let workspacePaths: string[] = [];
+	private async initialize(workspaceFolders: vscode.WorkspaceFolder[], bookPath: string): Promise<void> {
 		if (bookPath) {
-			workspacePaths.push(bookPath);
+			let book: BookModel = new BookModel(bookPath, this._openAsUntitled, this._extensionContext);
+			await book.initializeContents();
+			this.books.push(book);
+			if (!this.currentBook) {
+				this.currentBook = book;
+			}
 		}
 		else if (workspaceFolders) {
-			workspacePaths = workspaceFolders.map(a => a.uri.fsPath);
+			workspaceFolders.map(async (a) => {
+				let book: BookModel = new BookModel(a.uri.fsPath, this._openAsUntitled, this._extensionContext);
+				await book.initializeContents();
+				this.books.push(book);
+				if (!this.currentBook) {
+					this.currentBook = book;
+				}
+			});
 		}
-		this.getTableOfContentFiles(workspacePaths).then(() => undefined, (err) => { console.log(err); });
-		this._extensionContext = context;
 	}
 
 	public get onReadAllTOCFiles(): vscode.Event<void> {
 		return this._onReadAllTOCFiles.event;
 	}
 
-	async getTableOfContentFiles(workspacePaths: string[]): Promise<void> {
-		let notebookConfig = vscode.workspace.getConfiguration(notebookConfigKey);
-		let maxDepth = notebookConfig[maxBookSearchDepth];
-		// Use default value if user enters an invalid value
-		if (maxDepth === undefined || maxDepth < 0) {
-			maxDepth = 5;
-		} else if (maxDepth === 0) { // No limit of search depth if user enters 0
-			maxDepth = undefined;
-		}
-		for (let workspacePath of workspacePaths) {
-			let p = path.join(workspacePath, '**', '_data', 'toc.yml').replace(/\\/g, '/');
-			let tableOfContentPaths = await glob(p, { deep: maxDepth });
-			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
-		}
-		let bookOpened: boolean = this._tableOfContentPaths.length > 0;
-		vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
-		this._onReadAllTOCFiles.fire();
-	}
 
-	async openBook(bookPath: string, openAsUntitled: boolean, urlToOpen?: string): Promise<void> {
+
+	async openBook(bookPath: string, urlToOpen?: string): Promise<void> {
 		try {
+			let book: BookModel = this.books.filter(book => book.BookPath === bookPath)[0];
 			// Check if the book is already open in viewlet.
```

Comment:
You should be checking that there's at least one returned value - this will break if passed an unknown path